### PR TITLE
Replay logging integration + replay backend

### LINKED ISSUE #1584: Capture replay logging test data with Refbox messages
### Description of the task

Capture replay logging test data with Refbox messages, validate that the messages created by `ReplayReader` are well formed in Python, and then replace the current test data with the newly created test data

### Acceptance criteria

- [ ] logged protobuf data with SensorMsg's containing one or more `Referee` messages
- [ ] data can be opened and is well-formed when deserialized in Python
- [ ] add to replay_logging unit tests

### LINKED ISSUE #1624: Create a backend for replaying from logged messages
### Acceptance criteria

- [ ] Replay backend

### LINKED ISSUE #1625: Integrate replay logging
### Acceptance criteria

- [ ] SensorMsgs are logged during normal operation
- [ ] Dynamic parameters to select whether to log, and what directory to log to

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description

- Integrate replay logging into `full_system_main`. Replays will log to a directory given in the command-line argument `--replay_output_dir`, or no logging will happen if this argument is not set.
- Create `ReplayBackend` which takes the command-line argument `--replay_input_dir` and replays the logged data at real time. 

### Testing Done

- Works on my machine
- Updated unit tests for replay to have `Referee` messages

### Resolved Issues

Resolves  #1625, resolves #1624, resolves #1584

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/backend/replay_backend.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        this->sendValueToObservers(*sensor_msg_or_null);
+        last_msg_replayed_time = std::chrono::steady_clock::now();
+        last_msg_received_time = this_msg_received_time;
+    }
+    throw std::out_of_range("Reached end of replay!");
+}
+
+// Register this backend in the genericFactory
+static TGenericFactory<std::string, Backend, ReplayBackend> factory;
```
Comment (kelvinkoon) [inline src/software/backend/replay_backend.cpp]: Nit: EOF

File: src/software/backend/replay_backend.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#include "replay_backend.h"
+
+#include "software/util/design_patterns/generic_factory.h"
+
+ReplayBackend::ReplayBackend(const std::string& replay_input_dir)
+    : Backend(),
+      replay_reader(replay_input_dir),
```
Comment (jonathanlew) [inline src/software/backend/replay_backend.cpp]: does this fail nicely when the `replay_input_dir` isn't valid?

File: src/software/backend/replay_logging/replay_test.cpp [hunk 3]
```diff
@@ -39,17 +39,20 @@ TEST(ReplayTest, test_read_and_write_replay)
     // do a quick sanity check here to assert that the timestamps are monotonically
     // increasing
     double max_timestamp = 0.f;
+    size_t idx           = 0;
```
Comment (kelvinkoon) [inline src/software/backend/replay_logging/replay_test.cpp]: What's the purpose of `idx` exactly? It isn't used to verify, so is it strictly for debugging?

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Looks good, just checking a few things

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
This is awesome! Just ran it locally and the replay worked great

Just two things I noticed:
* We need to add a check for if `--replay_input_dir` is provided, the `--backend` is set to `ReplayBackend`. If that's not the case, `full_system_main` should complain
* There is an `std::out_of_range` thrown at the end of replay, it doesn't effect the functionality, but it would be nice if the backend elegantly shut down

Nice to have:
* Currently `--replay_output_dir` requires you to create a path and give that to the replay backend, how easy would it be to make the directory automatically if it doesn't exist? Just for convenience sake

**Review by jonathanlew — COMMENTED** [review]
Just wondering if we can we run this with the simulator

File: src/software/backend/replay_backend.cpp (hunk 2 above)
Comment (jordanzliu) [reply src/software/backend/replay_backend.cpp]: `ReplayReader` will throw

File: src/software/backend/replay_logging/replay_test.cpp (hunk 3 above)
Comment (jordanzliu) [reply src/software/backend/replay_logging/replay_test.cpp]: yeah it's for debugging, perhaps this isn't necessary but this was a nice-to-have for when I was debugging using python protobuf

**Comment by jordanzliu** [comment]
@akhilveeraghanta 
- `ReplayReader` will throw if the input path is an invalid path, this includes empty string. Since the default of the `--replay_input_dir` is empty string, it will complain if the argument is not provided at all.
- Fixed exit behaviour to, after we run out of replay messagse, wait 1 second after the last primitive was received by the backend
- It does currently create a directory, but if the given path's parent directory does not exist, it will throw

File: src/software/backend/replay_backend.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      pull_from_replay_thread(
+          boost::bind(&ReplayBackend::continuouslyPullFromReplayFiles, this)),
+      last_msg_received_time(std::nullopt),
+      last_msg_replayed_time(std::nullopt)
+{
+}
+
+// do nothing
+// TODO: #1745 perhaps we want to encode the output primitives to proto and log them?
```
Comment (kelvinkoon) [inline src/software/backend/replay_backend.cpp]: Putting the issue link as well would be nice to keep it consistent with the rest of the TODOs.

Comment (jonathanlew) [reply src/software/backend/replay_backend.cpp]: issue link is not required; either way is fine

File: src/software/backend/backend.cpp [hunk 5]
```diff
@@ -1,22 +1,27 @@
 #include "software/backend/backend.h"
 
+#include "software/proto/message_translation/tbots_protobuf.h"
+
 void Backend::receiveRobotStatus(TbotsProto::RobotStatus msg)
 {
     SensorProto sensor_msg;
     *(sensor_msg.add_robot_status_msgs()) = msg;
+    (*sensor_msg.mutable_time_received()) = *createCurrentTimestamp();
```
Comment (jonathanlew) [inline src/software/backend/backend.cpp]: nit: looks kinda weird, but it probably works. Same for the rest of the file
```suggestion
    *(sensor_msg.mutable_time_received()) = *createCurrentTimestamp();
```

File: src/software/proto/sensor_msg.proto [hunk 6]
```diff
@@ -9,4 +10,6 @@ message SensorProto
     SSLProto.Referee ssl_referee_msg                  = 1;
     SSLProto.SSL_WrapperPacket ssl_vision_msg         = 2;
     repeated TbotsProto.RobotStatus robot_status_msgs = 3;
+    // this is only used for replay at the moment
+    TbotsProto.Timestamp time_received = 4;
```
Comment (jonathanlew) [inline src/software/proto/sensor_msg.proto]: the name is ambiguous here, so how about
```suggestion
    // this is only used for replay at the moment
    TbotsProto.Timestamp backend_receive_time = 4;
```

File: src/software/backend/replay_logging/replay_test.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/backend/replay_logging/replay_test.cpp]: please add a comment about it then

File: src/software/backend/replay_backend.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
+#include "replay_backend.h"
+
+#include <cstdlib>
+
+#include "software/util/design_patterns/generic_factory.h"
+
+ReplayBackend::ReplayBackend(const std::string& replay_input_dir)
+    : Backend(),
```
Comment (jonathanlew) [inline src/software/backend/replay_backend.cpp]: nit: I don't think the `Backend` constructor needs to be called

File: src/software/backend/replay_backend.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        auto this_msg_received_time = std::chrono::duration<double>(
+            sensor_msg_or_null->time_received().epoch_timestamp_seconds());
+
+        if (last_msg_received_time && last_msg_replayed_time)
+        {
+            std::this_thread::sleep_until(
+                *last_msg_replayed_time +
+                (this_msg_received_time - *last_msg_received_time));
+        }
```
Comment (jonathanlew) [inline src/software/backend/replay_backend.cpp]: can we add a quick comment to explain that we want to replicate the timing of messages by sleeping

File: src/software/backend/replay_backend.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        this->sendValueToObservers(*sensor_msg_or_null);
+        last_msg_replayed_time = std::chrono::steady_clock::now();
+        last_msg_received_time = this_msg_received_time;
+    }
+
+    // wait 1 second until the last primitive is received by the backend
+    // to exit
+    bool exit = false;
```
Comment (jonathanlew) [inline src/software/backend/replay_backend.cpp]: nit: we don't really need this comment given the comment below

**Review by jonathanlew — CHANGES_REQUESTED** [review]
just some minor comments

File: src/software/backend/replay_backend.cpp (hunk 7 above)
Comment (jordanzliu) [reply src/software/backend/replay_backend.cpp]: removed default constructors here

**Review by jonathanlew — COMMENTED** [review]
LGTM!

**Review by jonathanlew — COMMENTED** [review]
Still LGTM waiting on Akhil and Kelvin's reviews

**Review by kelvinkoon — APPROVED** [review]
:rewind:

**Review by akhilveeraghanta — APPROVED** [review]
Tried it out locally and everything works great!

LGTM :100:

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**